### PR TITLE
[SPARK-37279][PYTHON][SQL] Support DayTimeIntervalType in createDataFrame, collect and Python UDF

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR implements `DayTimeIntervalType` in PySpark's `DataFrame.collect()`, `SparkSession.createDataFrame()` and `functions.udf`.
This type is mapped to [`datetime.timedelta`](https://docs.python.org/3/library/datetime.html#timedelta-objects).

Arrow code path will be separately implemented at SPARK-37277, and Py4J support will be done at SPARK-37281.

### Why are the changes needed?

- In order to support `datetime.timedelta` out of the box via PySpark.
- To seamlessly support ANSI standard types

Semantically [`datetime.timedelta`](https://docs.python.org/3/library/datetime.html#timedelta-objects) is mapped to `DayTimeIntervalType`. Python's timedelta does not support months, years, etc.

### Does this PR introduce _any_ user-facing change?

Yes, users will be able to use `datetime.timedelta` in PySpark with `DayTimeIntervalType` at `DataFrame.collect()`, `SparkSession.createDataFrame()` and `functions.udf`:

```python
>>> import datetime
>>> df = spark.createDataFrame([(datetime.timedelta(days=1),)])
>>> df.collect()
[Row(_1=datetime.timedelta(days=1))]
```

```python
>>> from pyspark.sql.functions import udf
>>> df.select(udf(lambda x: x, "interval day to second")("_1")).show()
+--------------------+
|        <lambda>(_1)|
+--------------------+
|INTERVAL '1 00:00...|
+--------------------+
```

### How was this patch tested?

Unittests were added, and the

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -1034,11 +1089,19 @@ def _parse_datatype_json_value(json_value: Union[dict, str]) -> DataType:
             return _all_atomic_types[json_value]()
         elif json_value == "decimal":
             return DecimalType()
-        elif json_value == "timestamp_ntz":
-            return TimestampNTZType()
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:1038]: This is unreachable. `if json_value in _all_atomic_types.keys():` above handles it.

**Comment by HyukjinKwon** [comment]
cc @ueshin @MaxGekk @itholic @xinrong-databricks FYI

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
This is ready for a look.

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -317,6 +319,65 @@ def simpleString(self) -> str:
...
+
+    jsonValue = _str_repr
+
+    def __repr__(self) -> str:
+        return "%s(%d,%d)" % (type(self).__name__, self.startField, self.endField)
+
+    def needConversion(self) -> bool:
+        return True
+
+    def toInternal(self, dt: datetime.timedelta) -> Optional[int]:
```
Comment (cloud-fan) [inline python/pyspark/sql/types.py:372]: internally day-time interval is stored as long, why do we use int here?

File: python/pyspark/sql/types.py [hunk 3]
```diff
@@ -317,6 +319,65 @@ def simpleString(self) -> str:
...
+        if dt is not None:
+            return (math.floor(dt.total_seconds()) * 1000000) + dt.microseconds
+
+    def fromInternal(self, micros: int) -> Optional[datetime.timedelta]:
```
Comment (cloud-fan) [inline python/pyspark/sql/types.py:376]: ditto

**Comment by cloud-fan** [comment]
I'm not very familiar with PySpark. How do we support `spark.createDataFrame`? I can't find related code in this PR.

File: python/pyspark/sql/types.py (hunk 2 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/types.py:372]: Python's int is actually long .. from Python 3 :-).

Comment (HyukjinKwon) [reply python/pyspark/sql/types.py:372]: and this part of codes are shared for `createDataFrame`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala [hunk 4]
```diff
@@ -137,11 +137,12 @@ object EvaluatePython {
       case c: Int => c
     }
 
-    case TimestampType | TimestampNTZType => (obj: Any) => nullSafeConvert(obj) {
-      case c: Long => c
-      // Py4J serializes values between MIN_INT and MAX_INT as Ints, not Longs
-      case c: Int => c.toLong
-    }
+    case TimestampType | TimestampNTZType | _: DayTimeIntervalType => (obj: Any) =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala:140]: e.g.) PySpark reads internal data format of `DayTimeIntervalType` as `Long`, and convert it to Python instances at https://github.com/apache/spark/pull/34614/files#diff-bb02965370d515ec7add7967bc949c18c88ffc7346cceaebefb84907dfdce903R376 for both createDataFrame and Python UDF

**Review by cloud-fan — APPROVED** [review]

File: python/pyspark/sql/types.py [hunk 5]
```diff
@@ -1034,11 +1096,17 @@ def _parse_datatype_json_value(json_value: Union[dict, str]) -> DataType:
...
         elif json_value == "decimal":
             return DecimalType()
-        elif json_value == "timestamp_ntz":
-            return TimestampNTZType()
         elif _FIXED_DECIMAL.match(json_value):
             m = _FIXED_DECIMAL.match(json_value)
             return DecimalType(int(m.group(1)), int(m.group(2)))  # type: ignore[union-attr]
+        elif _INTERVAL_DAYTIME.match(json_value):
+            m = _INTERVAL_DAYTIME.match(json_value)
+            inverted_fields = DayTimeIntervalType._inverted_fields  # type: ignore[union-attr]
```
Comment (ueshin) [inline python/pyspark/sql/types.py:1104]: nit: we don't need the ignore here.

**Review by ueshin — APPROVED** [review]
LGTM.

File: python/pyspark/sql/types.py (hunk 5 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:1104]:
```suggestion
            inverted_fields = DayTimeIntervalType._inverted_fields
```

**Comment by HyukjinKwon** [comment]
Last commit was just removing a comment.

Merged to master.

**Pull request closed without merging**